Use new activerecord 4 'references' feature

movitto opened the pull request:
Fixes error occurring during startup sequence

```
/opt/rubies/ruby-2.0.0-p645/lib/ruby/gems/2.0.0/bundler/gems/rails-07c6433ba9e2/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:596:in `exec': PG::Error: ERROR:  missing FROM-clause entry for table "guest_devices" (ActiveRecord::StatementInvalid)
LINE 1: ...ERE ((hardwares.vm_or_template_id IS NOT NULL AND guest_devi...
                                                             ^
: SELECT "hardwares".* FROM "hardwares" WHERE ((hardwares.vm_or_template_id IS NOT NULL AND guest_devices.address = '52:54:00:45:FD:A4' AND networks.hostname = 'localhost.localdomain.localdomain' AND networks.ipaddress = '192.168.122.140'))

	from /var/www/miq/vmdb/app/models/vm.rb:47:in `find_all_by_mac_address_and_hostname_and_ipaddress'
	from /var/www/miq/vmdb/app/models/miq_server.rb:237:in `start'
	from /var/www/miq/vmdb/lib/workers/evm_server.rb:68:in `start'
```

chessbyte commented:
@matthewd @tenderlove @Fryguy Please review

jrafanie commented:
@movitto, thanks... I'm writing a test for this change.

jrafanie commented:
@movitto Please add/cleanup the following patch... It recreates the issue and verifies your fix:

```
From 75eef439c4e4932f0a17da4a5e2fc6111a05dd4c Mon Sep 17 00:00:00 2001
From: Joe Rafaniello <jrafanie@redhat.com>
Date: Tue, 16 Jun 2015 10:17:27 -0400
Subject: [PATCH] Add test for just mac_address case.

---
 vmdb/spec/models/vm_spec.rb | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/vmdb/spec/models/vm_spec.rb b/vmdb/spec/models/vm_spec.rb
index b3dacc4..2aade8e 100644
--- a/vmdb/spec/models/vm_spec.rb
+++ b/vmdb/spec/models/vm_spec.rb
@@ -40,6 +40,25 @@ describe Vm do
     lambda { vm.validate_remote_console_vmrc_support }.should raise_error MiqException::RemoteConsoleNotSupportedError
   end

+  context ".find_all_by_mac_address_and_hostname_and_ipaddress" do
+    before do
+      @hardware1 = FactoryGirl.create(:hardware)
+      @vm1 = FactoryGirl.create(:vm_vmware, :hardware => @hardware1)
+
+      @hardware2 = FactoryGirl.create(:hardware)
+      @vm2 = FactoryGirl.create(:vm_vmware, :hardware => @hardware2)
+    end
+
+    it "mac_address" do
+      address = "ABCDEFG"
+      guest_device = FactoryGirl.create(:guest_device, :address => address, :device_type => "ethernet")
+      @hardware1.guest_devices << guest_device
+
+      expect(described_class.find_all_by_mac_address_and_hostname_and_ipaddress(address, nil, nil))
+        .to eql([@vm1])
+    end
+  end
+
   context "with relationships of multiple types" do
     before(:each) do
       @rp        = FactoryGirl.create(:resource_pool, :name => "RP")
--
2.2.1
```

jrafanie commented:
cc @lfu @tenderlove (this is the same issue you were chatting about in irc yesterday.

lfu commented:
@jrafanie That is right. I made PR #3181 that fixes the same issue.
cc @tenderlove

Fryguy commented:
Let's fix here since @jrafanie supplied a test.

tenderlove commented:
Ya, with the test, this PR looks fine.

matthewd commented:
We need to combine this with #3181 -- the `networks` branches here need to be adding to `references` too.

movitto commented:
@jrafanie Verified the test and added it to the PR

Note I left the "includes(include)" as is and added "guest_devices" to a new array which gets passed onto "references". Where as in lucy's pr she passes "include" onto "references". If this is the correct correct this pr needs to be updated.

jrafanie commented:
Thanks @movitto, can you also do the other legs of the condition to do the same with networks as @matthewd mentioned above?  You should be able to add tests for `if ipaddress` and `if hostname` just like the one I added for mac address.

Thanks for fixing this!

matthewd commented:
Yeah.. we need to reference `:networks` where we use it, but `:guest_devices` (as opposed to `:nics`) is the right one for the top conditional. (I have Opinions on whether that's a good API choice, but it's what the method is currently documented as wanting.)

jrafanie commented:
@movitto Please rebase to pull in the fix in #3204 to fix your test failure.  Thanks!

movitto commented:
@jrafanie rebased

jrafanie commented:
Thanks @movitto
